src/definitions/utils.d.ts

```diff
@@ -0,0 +1,19 @@
+type GetOptional<T> = {
+  [K in keyof T as Pick<T, K> extends Required<Pick<T, K>> ? never : K]: T[K];
+};
+
+type GetRequired<T> = {
+  [K in keyof T as Pick<T, K> extends Required<Pick<T, K>> ? K : never]: T[K];
+};
+
+type UnionUndefined<T> = {
+  [K in keyof T]: T[K] | undefined;
+};
```

Comment:
Did we want to expose these globally?

Reply:
Personally, I've never really bothered thinking about that. If that's an issue we could just change it to a manual import, or I could use a namespace for the utils - whichever you prefer.

With the namespace way `Utils` would be globally available but it does clutter it less, we could put all the types in the `Utils` namespace so the only "globally" available type is the `Utils` variable.

```ts
declare namespace Utils {
  type GetOptional ...
}
type EnforceUndefined = Utils.GetOptional...

// some other file
EnforcedUndefined<... // This is OK
GetOptional<... // would error
Utils.GetOptional<... // This is also OK
```

Reply:
@pokey What approach do you want me to take here? `import { EnforceUndefined } from '...'` or the Utils namespace or just leave it as is?

Reply:
Tbh I'd be tempted to just require users to import `EnforceUndefined`.  With auto-import, that's not particularly onerous, and it makes the code a bit better self-documenting, as you can see in the imports where it's coming from.  I'm not a huge fan of globally available symbols